juju/juju
# chore: remove call to legacy Charm() on bootstrap

In order to start cleaning the legacy state out of Charm() calls, we need to first remove the one on the bootstrap worker so that Juju is able to deploy the controller charm without retrieving the charm from the legacy state.

This is possible because both for local and charmhub charms we have everything we need (namely URL and Meta), and therefore the only big change in this patch is to add the CharmURL to the attributes on AddApplication() and to move the URL() method out from CharmRef to CharmRefFull, this way we can pass everything during bootstrap.





## QA steps

Since the `AddApplication()` method has changed, with a new attribute, we must test bootstrapping and deploying (both local / ch charms):
```
$ juju bootstrap lxd c
$ juju download ubuntu
$ juju deploy ./ubuntu_r25.charm local
$ juju deploy ubuntu ch
```
Make sure there are no errors on the logs.

Also bootstrap with a local controller charm:
```
$ juju download juju-controller
Base "ubuntu@24.04" is not supported for charm "juju-controller", trying base "ubuntu@22.04"
Fetching charm "juju-controller" revision 91 using "stable" channel and base "amd64/ubuntu/22.04"
Install the "juju-controller" charm with:
    juju deploy ./juju-controller_r91.charm
$ juju bootstrap localhost c --build-agent --controller-charm-path ./juju-controller_r91.charm
```

Check if the controller is OK, simply with `juju status -m controller`.
## Links


**Jira card:** JUJU-7256
